<a href="https://colab.research.google.com/github/vperls000/csc44800/blob/main/Viviana_Perls_Assignment3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/vperls000/csc44800/blob/main/Viviana_Perls_Assignment3.ipynb)

In [ ]:
# Imports section
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
import pandas as pd
import numpy as np

## Part 1. Loading the dataset

In [ ]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
slime = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
slime.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)
slime.describe()

Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

## Part 2. Splitting the dataset

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)

Y = slime['Size nm^3']
X = slime[['Temperature °C','Mols KCL']]
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=.1, train_size=.9, random_state=0, shuffle=True, stratify=None)

## Part 3. Perform a Linear Regression

In [ ]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)


# Create a sample datapoint and predict the output of that sample with the trained model
X_sample = [[750.250000, 710.250000]]
X_pred = pd.DataFrame(X_sample, columns=X_test.columns)
y_pred = reg.predict(X_pred)
print(y_pred)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print('Score: ', reg.score(X_test, y_test))
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print('Coefficients: ', reg.coef_)
print('Intercept: ', reg.intercept_)

[962197.79736605]
Score:  0.8761646752736477
Coefficients:  [ 863.58108791 1006.12741921]
Intercept:  -400305.9133335322


The score (about 87.6%) is not perfect, which is further shown by the predicted size of my X_sample.  The feature values for X_sample came from the 75th percentile of the original slime dataset, and the size is approximately two cubic nanometers less than the predicted size.  The score essentially tells us what percentage of test samples were correctly classified.

$h(x_1,x_2)=863.58x_1+1006.1x_2-400310$

## Part 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(reg, X, Y, cv=10)
print(scores)
# Report on their finding and their significance

[0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]


I shuffled the data 10 times to see if it would be very consistent or not.  The minimum score was about 78.3% and the maximum score was about 88.6%.  Since the maximum score is higher than the score from the first run, I assume that the training data was better in the former.

## Part 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
X_pt5 = PolynomialFeatures(degree=2).fit_transform(X)
X_pt5_train, X_pt5_test, y_pt5_train, y_pt5_test = train_test_split(X_pt5, Y, test_size=.1, train_size=.9, random_state=0, shuffle=True, stratify=None)
polyreg = LinearRegression().fit(X_pt5_train, y_pt5_train)
# Report on the metrics and output the resultant equation as you did in Part 3.
print('Score: ', polyreg.score(X_pt5_test, y_pt5_test))
print('Coefficients: ', polyreg.coef_)
print('Intercept: ', polyreg.intercept_)

Score:  1.0
Coefficients:  [ 0.00000000e+00  1.20000000e+01 -1.10383439e-07 -1.20916610e-11
  2.00000000e+00  2.85714287e-02]
Intercept:  1.3900746125727892e-05


Based on the coefficients shown above, the weights of $x_1$ and $x_1^2$ were higher, therefore the Mols KCL were more relevant to determining size than the Temperature (C).  The score of one implies that this model is much more effective than the linear regression model.

$h(x)=1.3901*10^{-5}+1.2000*10x_0-1.1038*10^{-7}x_1-1.2091*10^{-11}x_0^2+2.0000x_0x_1+2.8571x_1^2$